### PR TITLE
Add option to change material color from ROS.

### PR DESCRIPTION
**Pull request by bperseghetti**
# 🎉 New feature

## Summary
Allows for [setting a GZ entities visual material color](https://github.com/gazebosim/gz-sim/pull/2286) from ROS. This is highly valuable for creating noticeable LED status lighting. And depends on associated [gz-msgs](https://github.com/gazebosim/gz-msgs/pull/414).
## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [x] Updated documentation (as needed)
- [x] Updated migration guide (as needed)
- [x] Consider updating Python bindings (if the library has them)
- [x] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [x] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: ros_gz_interfaces/msg/MaterialColor.msg [hunk 1]
```diff
@@ -0,0 +1,9 @@
+std_msgs/Header header            # Optional header data
+
+string name                		  # name
+string parent_name                # parent name
+
+std_msgs/ColorRGBA ambient        # ambient
```
Comment (ahcorde) [inline ros_gz_interfaces/msg/MaterialColor.msg]:
```suggestion

string name                	  # name
string parent_name                # parent name

std_msgs/ColorRGBA ambient        # ambient 
```

**Review by ahcorde — CHANGES_REQUESTED** [review]
can you add a test?

**Comment by bperseghetti** [comment]
> can you add a test?

Yes, I was planning to wait and make sure this change was desired in upstream before spending any extra time to add a test.
Also to make testing work properly with CI will need to get https://github.com/gazebosim/gz-msgs/pull/414 in first anyhow. This was more just to show the interdependent changes.

**Comment by bperseghetti** [comment]
@ahcorde Test added.

**Review by ahcorde — CHANGES_REQUESTED** [review]
you included some changes related with renaming `ign` to `gz`. Do you mind to remove them?

**Comment by bperseghetti** [comment]
> you included some changes related with renaming `ign` to `gz`. Do you mind to remove them?

So that readme is in the ros_gz_bridge instead of the ros_ign_bridge, should it not be changed to match the naming?

**Comment by azeey** [comment]
> So that readme is in the ros_gz_bridge instead of the ros_ign_bridge, should it not be changed to match the naming?

I think we should stick with `ign` since Humble is officially paired with Fortress. It's confusing for users since trying to use `gz` doesn't work in Humble/Fortress.

**Comment by bperseghetti** [comment]
> > So that readme is in the ros_gz_bridge instead of the ros_ign_bridge, should it not be changed to match the naming?
> 
> I think we should stick with `ign` since Humble is officially paired with Fortress. It's confusing for users since trying to use `gz` doesn't work in Humble/Fortress.

But to be clear this is not the main README but this is the one inside ros_gz_bridge. The one inside ros_ign_bridge should just say it's a shim on the other and have ign listed in it, right?

I guess I should just swap them.

**Comment by azeey** [comment]
> But to be clear this is not the main README but this is the one inside ros_gz_bridge. The one inside ros_ign_bridge should just say it's a shim on the other and have ign listed in it, right?

https://github.com/gazebosim/ros_gz/tree/humble/ros_ign_bridge does say it's a shim

**Comment by bperseghetti** [comment]
> > But to be clear this is not the main README but this is the one inside ros_gz_bridge. The one inside ros_ign_bridge should just say it's a shim on the other and have ign listed in it, right?
> 
> https://github.com/gazebosim/ros_gz/tree/humble/ros_ign_bridge does say it's a shim

Well I did also just change that so you know it's a shim but it has different interface and that gives you the interface of it.

**Comment by bperseghetti** [comment]
@osrf-jenkins run tests please

**Comment by azeey** [comment]
This PR is targeting `humble` which needs to work with Fortress, but the gz-msgs and gz-sim that landed recently all targeted Harmonic. I recommend retargeting this to `rolling`.

**Comment by azeey** [comment]
Alternatively, you can do something like https://github.com/gazebosim/ros_gz/blob/84d5d27e6f730db506e6b229c40ef3fdc7042cf6/ros_gz_bridge/include/ros_gz_bridge/ros_gz_bridge.hpp#L26-L38

**Comment by bperseghetti** [comment]
> Alternatively, you can do something like
> 
> https://github.com/gazebosim/ros_gz/blob/84d5d27e6f730db506e6b229c40ef3fdc7042cf6/ros_gz_bridge/include/ros_gz_bridge/ros_gz_bridge.hpp#L26-L38

Yep, can't believe I didn't think of/see that. Thanks!

**Comment by bperseghetti** [comment]
> This PR is targeting `humble` which needs to work with Fortress, but the gz-msgs and gz-sim that landed recently all targeted Harmonic. I recommend retargeting this to `rolling`.

I do want this to land in humble first and foremost as that is the "LTS", I can also bump it to rolling in another eventual PR and at that point remove the if/def conditionals from the code.

File: ros_gz_bridge/README.md [hunk 2]
```diff
@@ -5,70 +5,71 @@ between ROS and Gazebo Transport.
...
-| sensor_msgs/msg/Image                       | ignition::msgs::Image                  |
-| sensor_msgs/msg/JointState                  | ignition::msgs::Model                  |
-| sensor_msgs/msg/Joy                         | ignition::msgs::Joy                    |
-| sensor_msgs/msg/LaserScan                   | ignition::msgs::LaserScan              |
-| sensor_msgs/msg/MagneticField               | ignition::msgs::Magnetometer           |
-| sensor_msgs/msg/NavSatFix                   | ignition::msgs::NavSat                 |
-| sensor_msgs/msg/PointCloud2                 | ignition::msgs::PointCloudPacked       |
-| tf2_msgs/msg/TFMessage                      | ignition::msgs::Pose_V                 |
-| trajectory_msgs/msg/JointTrajectory         | ignition::msgs::JointTrajectory        |
+| ROS type                             | Gazebo type                            |
```
Comment (azeey) [inline ros_gz_bridge/README.md]: Please revert the changes in this file. I know the directory is `ros_gz_bridge`, but the example commands will not work when using Fortress, which is the default in Humble.

File: ros_gz_bridge/test/utils/gz_test_msg.cpp [hunk 3]
```diff
@@ -1346,6 +1346,50 @@ void compareTestMsg(const std::shared_ptr<gz::msgs::Light> & _msg)
...
+  createTestMsg(diffuse_msg);
+  createTestMsg(specular_msg);
+  createTestMsg(emissive_msg);
+
+  _msg.mutable_header()->CopyFrom(header_msg);
+  _msg.mutable_entity()->CopyFrom(entity_msg);
+  _msg.mutable_specular()->CopyFrom(ambient_msg);
+  _msg.mutable_diffuse()->CopyFrom(diffuse_msg);
+  _msg.mutable_specular()->CopyFrom(specular_msg);
+  _msg.mutable_specular()->CopyFrom(emissive_msg);
```
Comment (azeey) [inline ros_gz_bridge/test/utils/gz_test_msg.cpp]: There's a copy/paste error here. `mutable_specular` is used 3 times.

Comment (azeey) [inline ros_gz_bridge/test/utils/gz_test_msg.cpp]: I would think this would work instead of creating temporaries. 
```c++
  createTestMsg(*_msg.mutable_header());
  createTestMsg(*_msg.mutable_entity());
  createTestMsg(*_msg.mutable_ambient());
  createTestMsg(*_msg.mutable_diffuse());
  createTestMsg(*_msg.mutable_specular());
  createTestMsg(*_msg.mutable_emissive());
```

File: ros_ign_bridge/README.md [hunk 4]
```diff
@@ -1,2 +1,284 @@
 # This is a shim package
 For [ros_gz_bridge](https://github.com/gazebosim/ros_gz/tree/ros2/ros_gz_bridge)
+
```
Comment (azeey) [inline ros_ign_bridge/README.md]: I'd rather not maintain two README files. Can we just keep the one in `ros_gz_bridge`?

**Review by azeey — CHANGES_REQUESTED** [review]

File: ros_gz_bridge/test/utils/gz_test_msg.cpp (hunk 3 above)
Comment (bperseghetti) [reply ros_gz_bridge/test/utils/gz_test_msg.cpp]: Whoooops, yeah, must have been code blind there. I don't know how many times I stared right at that and was like "hmmm, nothing seems wrong" 🤦🏼‍♂️

Comment (bperseghetti) [reply ros_gz_bridge/test/utils/gz_test_msg.cpp]: Yeah, makes sense!

File: ros_ign_bridge/README.md (hunk 4 above)
Comment (bperseghetti) [reply ros_ign_bridge/README.md]: Yeah, I don't care about updating the README anymore lol.

File: ros_gz_bridge/README.md (hunk 2 above)
Comment (bperseghetti) [reply ros_gz_bridge/README.md]: Yep, reverted.

**Review by ahcorde — CHANGES_REQUESTED** [review]
linters
```bash
"ros_gz_bridge" "--output-file" "/github/home/ws/build/ros_gz_bridge/ament_flake8/flake8.txt" "--command" "/opt/ros/rolling/bin/ament_flake8" "--xunit-file" "/github/home/ws/build/ros_gz_bridge/test_results/ros_gz_bridge/flake8.xunit.xml"
2024-02-01T03:46:02.0622501Z 10: Test timeout computed to be: 60
2024-02-01T03:46:02.1111404Z 10: -- run_test.py: invoking following command in '/github/home/ws/src/ros_gz/ros_gz_bridge':
2024-02-01T03:46:02.1112566Z 10:  - /opt/ros/rolling/bin/ament_flake8 --xunit-file /github/home/ws/build/ros_gz_bridge/test_results/ros_gz_bridge/flake8.xunit.xml
2024-02-01T03:46:02.4909897Z 10: from ros_gz_bridge.mappings import MAPPINGS, MAPPINGS_8_4_0, MAPPINGS_10_1_0
2024-02-01T03:46:02.4911413Z 10: ^
2024-02-01T03:46:02.4913079Z 10: 1     I101 Imported names are in the wrong order. Should be MAPPINGS, MAPPINGS_10_1_0, MAPPINGS_8_4_0
2024-02-01T03:46:02.4914637Z 10: 
2024-02-01T03:46:02.4916618Z 10: ./ros_gz_bridge/__init__.py:19:1: I101 Imported names are in the wrong order. Should be MAPPINGS, MAPPINGS_10_1_0, MAPPINGS_8_4_0
2024-02-01T03:46:02.4918587Z 10: 
2024-02-01T03:46:02.4919517Z 10: 
2024-02-01T03:46:02.4921177Z 10: 5 files checked
2024-02-01T03:46:02.4922615Z 10: 1 errors
2024-02-01T03:46:02.4924329Z 10: 
2024-02-01T03:46:02.4926578Z 10: 'I'-type errors: 1
2024-02-01T03:46:02.4928273Z 10: 
2024-02-01T03:46:02.4929977Z 10: Checked files:
```

**Review by ahcorde — APPROVED** [review]
@azeey any other comment?

**Review by azeey — APPROVED** [review]

**Comment by azeey** [comment]
@bperseghetti could you please cherry-pick this into `iron` and `ros2` branches?

**Comment by bperseghetti** [comment]
> @bperseghetti could you please cherry-pick this into `iron` and `ros2` branches?

Yep, I can do it later today!

**Comment by ahcorde** [comment]
friendly ping @bperseghetti do you have any plans to forward port these changes?

**Comment by bperseghetti** [comment]
Sorry, my bad, I let this one slip through the cracks on my side, will create the PRs after my meetings today.

**Comment by bperseghetti** [comment]
> friendly ping @bperseghetti do you have any plans to forward port these changes?

Randomly remembered to do this, sorry about the dealy, please see associated "forward feature ports":
https://github.com/gazebosim/ros_gz/pull/520
https://github.com/gazebosim/ros_gz/pull/521
